libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { IPhase } from '../../api/CommandLineConfiguration';
+import { EnvironmentVariableNames } from '../../api/EnvironmentConfiguration';
+import type {
+  ICreateOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import { NullOperationRunner } from './NullOperationRunner';
+import { Operation } from './Operation';
+import { OperationStatus } from './OperationStatus';
+import { ShellOperationRunner } from './ShellOperationRunner';
+import {
+  formatCommand,
+  getCustomParameterValuesByPhase,
+  getDisplayName,
+  getScriptToRun
+} from './ShellOperationRunnerPlugin';
+
+export const PLUGIN_NAME: 'ShardedPhasedOperationPlugin' = 'ShardedPhasedOperationPlugin';
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStrings = {
+  SHARD_INDEX: '{shardIndex}',
+  SHARD_COUNT: '{shardCount}'
+} as const;
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const TemplateStringRegex = {
+  SHARD_INDEX: new RegExp(TemplateStrings.SHARD_INDEX, 'g'),
+  SHARD_COUNT: new RegExp(TemplateStrings.SHARD_COUNT, 'g')
+} as const;
+
+/**
+ * Phased command that shards a phase into multiple operations.
+ */
+export class ShardedPhasedOperationPlugin implements IPhasedCommandPlugin {
+  public apply(hooks: PhasedCommandHooks): void {
+    hooks.createOperations.tap(PLUGIN_NAME, spliceShards);
+  }
+}
+
+function spliceShards(existingOperations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
+  const { rushConfiguration } = context;
+
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project, settings: operationSettings } = operation;
+    if (phase && project && operationSettings?.sharding && !operation.runner) {
+      const { count: shards, shardScriptConfiguration } = operationSettings.sharding;
+
+      existingOperations.delete(operation);
+
+      const collatorNode: Operation = new Operation({
+        phase,
+        project,
+        settings: operationSettings
+      });
+      existingOperations.add(collatorNode);
```

Comment:
`collatorNode` is literally `operation` with its dependencies redefined; we could just reuse the existing operation and remove and relocate the dependencies.

To reduce the number of edges in the graph we could also inject a noop that keeps all the dependencies and that all the shards depend on, i.e.:

```
depA -\          /- shard 1 -\
depB -- > noop < -- shard 2 -- > collator (reused operation)
depC -/          \- shard 3 -/
```

Reply:
Updated in [42f63c2](https://github.com/microsoft/rushstack/pull/4652/commits/42f63c2d88fea68022699345055e59b1036c9789)